src/rgw/rgw_sal_store.h

```diff
@@ -0,0 +1,782 @@
+			  std::vector<std::unique_ptr<RGWRole>>& roles) = 0;
+    virtual std::unique_ptr<RGWOIDCProvider> get_oidc_provider() = 0;
+    virtual int get_oidc_providers(const DoutPrefixProvider *dpp,
+				   const std::string& tenant,
+				   std::vector<std::unique_ptr<RGWOIDCProvider>>& providers) = 0;
+    virtual std::unique_ptr<Writer> get_append_writer(const DoutPrefixProvider *dpp,
+				  optional_yield y,
+				  std::unique_ptr<rgw::sal::Object> _head_obj,
+				  const rgw_user& owner,
+				  const rgw_placement_rule *ptail_placement_rule,
+				  const std::string& unique_tag,
+				  uint64_t position,
+				  uint64_t *cur_accounted_size) = 0;
+    virtual std::unique_ptr<Writer> get_atomic_writer(const DoutPrefixProvider *dpp,
+				  optional_yield y,
+				  std::unique_ptr<rgw::sal::Object> _head_obj,
+				  const rgw_user& owner,
+				  const rgw_placement_rule *ptail_placement_rule,
+				  uint64_t olh_epoch,
+				  const std::string& unique_tag) = 0;
+
+    virtual const std::string& get_compression_type(const rgw_placement_rule& rule) = 0;
+    virtual bool valid_placement(const rgw_placement_rule& rule) = 0;
+
+    virtual void finalize(void) = 0;
+
+    virtual CephContext* ctx(void) = 0;
+
+    virtual const std::string& get_luarocks_path() const = 0;
+    virtual void set_luarocks_path(const std::string& path) = 0;
+};
+
+class StoreUser : public User {
+  protected:
+    RGWUserInfo info;
+    RGWObjVersionTracker objv_tracker;
+    Attrs attrs;
+
+  public:
+    StoreUser() : info() {}
+    StoreUser(const rgw_user& _u) : info() { info.user_id = _u; }
+    StoreUser(const RGWUserInfo& _i) : info(_i) {}
+    StoreUser(StoreUser& _o) = default;
+    virtual ~StoreUser() = default;
+
+    virtual std::unique_ptr<User> clone() = 0;
+    virtual int list_buckets(const DoutPrefixProvider* dpp,
+			     const std::string& marker, const std::string& end_marker,
+			     uint64_t max, bool need_stats, BucketList& buckets,
+			     optional_yield y) = 0;
+    virtual int create_bucket(const DoutPrefixProvider* dpp,
+                            const rgw_bucket& b,
+                            const std::string& zonegroup_id,
+                            rgw_placement_rule& placement_rule,
+                            std::string& swift_ver_location,
+                            const RGWQuotaInfo* pquota_info,
+                            const RGWAccessControlPolicy& policy,
+			    Attrs& attrs,
+                            RGWBucketInfo& info,
+                            obj_version& ep_objv,
+			    bool exclusive,
+			    bool obj_lock_enabled,
+			    bool* existed,
+			    req_info& req_info,
+			    std::unique_ptr<Bucket>* bucket,
+			    optional_yield y) = 0;
+
+    virtual std::string& get_display_name() { return info.display_name; }
+    const std::string& get_tenant() { return info.user_id.tenant; }
+    void set_tenant(std::string& _t) { info.user_id.tenant = _t; }
+    const std::string& get_ns() { return info.user_id.ns; }
```

Comment:
please use `override` for all of the virtuals that are implemented by these `Store*` classes. the compiler can help us keep these derived classes consistent with the bases